### ISSUE TITLE
UI: Management buttons in grain list

### ISSUE THREAD

**Issue by phildini** [sandstorm-io/sandstorm#364]
It would be awesome if I could see a trash can to delete grains from the overview page, rather than having to click into the grain.

**Comment by ocdtrekkie** [comment]
Yeah, this is on my wishlist. A redesign of the Sandstorm shell is coming soon. The file picker will need a whole menu to access things like rename, delete, sharing, etc.

**Comment by neynah** [comment]
This is a pretty old issue but one we should address since many people are asking for a grain delete button on the grain list page. I think this is a useful change that will make grain management more smooth & efficient. Currently, each grain must be individually opened and then deleted in the top nav.

My rough proposal is to have a checkbox column, clicking the checkbox in the column header will select all grains, clicking it again will deselect all grains. Users of course can also check the specific grains they'd want to delete. (Similar behaviour seen in Gmail for example.) Delete button is inactive until a grain is selected. I've also shifted the restore back-up button to be aligned with the delete button (grain count & total grain size should take precedence).

Pros: this is a pretty familiar UI and convenient for those that want to delete many grains at once.

Cons: there is little room for error since a user could delete all their grains (since the list goes on forever) and there isn't a way of recovering these grains. We don't have any grain archiving. Even though users can still filter to an extent using search, we don't have column sorting which I think would be the most useful in most cases. The cons are pretty bad so I'd love to hear any ideas on how to mitigate these.

![grains1](https://cloud.githubusercontent.com/assets/855341/13538531/0adc02de-e200-11e5-99eb-b6a9fb4006fc.png)

![grains2](https://cloud.githubusercontent.com/assets/855341/13538615/85ec9380-e200-11e5-878c-669c76554c6a.png)

Current:
![grains3](https://cloud.githubusercontent.com/assets/855341/13538720/074826ec-e201-11e5-9907-e2ee0694a858.png)

**Comment by ocdtrekkie** [comment]
@neynah Would it be reasonable to have the individual checkboxes, but no "check all" box? Then you avoid people accidentally deleting all their grains, while still making it drastically easier to delete multiple at a time.

**Comment by ocdtrekkie** [comment]
Note that Gmail doesn't have a select all checkbox anymore, as far as I can tell.

**Comment by neynah** [comment]
@ocdtrekkie It does seem like the better option to omit select all for now. Ability to shift-click multiple grains for selection would be pretty awesome BTW...

I'll leave "Select all" out for now unless @paulproteus/others can convince me otherwise before the day is over. :)

**Comment by ocdtrekkie** [comment]
I would also love multi-select to offer a download backup option, though I am guessing that's a fair bit more work to get right

**Comment by zarvox** [comment]
@neynah I like what you've drawn up a lot.  I might suggest adding a horizontal rule above/below the row of buttons (just two right now) to help separate the "button row" from the rest of the content visually.

@ocdtrekkie Bulk backup generation will be harder, especially since browsers usually don't like it when a single page starts a whole pile of downloads.  The alternative (building a zip-of-zips or something like that on the server and then sending it to the client) is worse in terms of usability, may have high latency to prepare, and may consume a large amount of server resources while being built.  We'll definitely have to think more about how to make that a satisfactory workflow - making a bunch of backups feels like a batch operation more so than an interactive one, so we should make the UX appropriate.

**Comment by neynah** [comment]
@zarvox I think adding a horizontal row / some visual separation is a good suggestion. I imagine we will also be adding further buttons in the "button row" in the future too.

**Comment by ocdtrekkie** [comment]
@zarvox Would it be possible to do, for now, a single-select-only 'download backup' from this list, that maybe greys out when you're multi-selecting? Even being able to download the backup without loading the grain would be really nice. Not just because of grain loading time, but I really don't want the grain's last activity date to change because I'm backing up a bunch of them.

**Comment by zarvox** [comment]
@ocdtrekkie The button would also have to disable when you've selected any grain that isn't owned by you.  But what you described should be more-or-less workable, and *should* avoid changing your last activity date.

**Comment by ar-jan** [comment]
Since there is no trash can for grains, asking for confirmation seems like a good idea. And then deleting multiple grains may also be OK? Perhaps something like:

```
You are about to delete the following grains:
Grain 1
Grain 2
....
Grain 10
and 23 more grains.

Deleting a grain cannot be undone. Do you want to proceed?
```
A trash can would also be very nice, by the way ;)

**Comment by zarvox** [comment]
@ar-jan You're quite right, deletion is scarily permanent.  Any action that destructive should prompt for confirmation.

Several people have suggested either a trash can, probably with the properties "trashcanned grains no longer start, but their data stays around".  I'd also consider an option to "archive" grains, which would hide them from your grainlist, but keep them around and operational if you or anyone else wanted to open them again.

You'd of course need some way to go back and un"delete"/unarchive grains though, which is a whole additional set of UI/workflows, so we haven't built this sort of thing yet.  Maybe this could be accomplished with a checkbox "show archived grains" or something like that.

**Comment by ocdtrekkie** [comment]
@zarvox I remain a strong proponent that tagging or folders should be supported by core Sandstorm, partially for this reason. Rather than making special cases just for archived or trashed grains, you could simply have trash and archive labels, and hide them from the main grain list by default.

**Comment by zarvox** [comment]
It's true that labels solve this in a general way, though how would Sandstorm know to hide those labels by default when you go to `/grain`?  That's more configuration UI that would need designing, along with UI for managing labels in general.

Another interesting point is that `trash` and `archive` happen to provide the additional hint that the grain is less likely to be accessed again, which is useful if you wanted to, say, compress the storage of grains that are less actively used, or move them out to higher-latency bulk storage to reduce operational costs while retaining performance.  Both of which are things that *would* need integration at the platform level.

That said, I am a huge fan of faceted search, and tags/folders wind up being another excellent facet that you can use to great effect.  We'll keep all these things in mind.

It's also possible that some day we'd want to make the grain list "just another grain" that gets a capability to list grains you have access to and their properties, and then *that* UI could be user-defined...but that's an idea whose time is a ways away yet. :)

**Comment by ocdtrekkie** [comment]
@zarvox Yeah, I know the idea of a replaceable shell or grain list is a possibility that's really exciting, but that doesn't mean tagging or folders shouldn't be built into core. Because if someone changes their shell or grain list, everything shouldn't end up dumped into one big pile. Tagging or folders need to be portable between alternative UIs one might chose to use, so the Sandstorm platform needs to be the holder of that metadata.

Trash also should probably have auto-deleting capability after 30 days since last active or what-have-you. So there's definitely special rules that might need to apply to certain default labels. But I'd much rather look at the platform supporting labels than making a lot of trash or archive-specific UI rather than a general sense of 'file location' and the UI to move things around in it.

**Comment by JamborJan** [comment]
One input from me: sharing of grains should be also possible for multiple grains. E.g. when I have a new project member I want to share all grains related to a project with him and not every grain one by one.

**Comment by neynah** [comment]
We've now implemented multi-select and grain trash in #1950 which addresses @phildini's concerns on this issue.
